File: backend/app/schemas/sharepoint.py
import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Segment(StrictModel):
    id: str
    location: str
    text: str


class Evidence(StrictModel):
    segment_id: str
    quote: str = Field(min_length=1, max_length=1500)


class Finding(StrictModel):
    title: str = Field(max_length=1000)
    owner: str | None
    deadline: str | None
    status: Literal["unknown", "pending", "in_progress", "completed", "blocked"]
    priority: Literal["unknown", "low", "medium", "high"]
    evidence: list[Evidence] = Field(min_length=1, max_length=8)

    @field_validator("deadline")
    @classmethod
    def valid_deadline(cls, value):
        if value is not None:
            from datetime import date
            if date.fromisoformat(value).isoformat() != value:
                raise ValueError("Use an unambiguous ISO date")
        return value


class ExpiryFinding(StrictModel):
    title: str = Field(max_length=1000)
    date: str = Field(max_length=20)
    category: Literal["expiry", "renewal", "effective", "warranty", "milestone", "deadline", "other"] = "expiry"
    responsible: str | None = None
    evidence: list[Evidence] = Field(min_length=1, max_length=8)

    @field_validator("date")
    @classmethod
    def valid_date(cls, value):
        if value is not None:
            from datetime import date
            if date.fromisoformat(value).isoformat() != value:
                raise ValueError("Use an unambiguous ISO date")
        return value


class CommercialFinding(StrictModel):
    description: str = Field(max_length=1000)
    amount: float | None = None
    currency: str | None = Field(default=None, max_length=10)
    payment_terms: str | None = Field(default=None, max_length=500)
    billing_frequency: Literal["one_time", "monthly", "quarterly", "annual", "milestone", "unknown"] = "unknown"
    evidence: list[Evidence] = Field(min_length=1, max_length=8)


class TextFact(StrictModel):
    value: str = Field(min_length=1, max_length=500)
    evidence: list[Evidence] = Field(min_length=1, max_length=5)


class DateFact(TextFact):
    @field_validator("value")
    @classmethod
    def valid_date(cls, value):
        from datetime import date
        if date.fromisoformat(value).isoformat() != value:
            raise ValueError("Use an unambiguous ISO date")
        return value


class NoticeFact(StrictModel):
    days: int = Field(ge=1, le=730)
    evidence: list[Evidence] = Field(min_length=1, max_length=5)


class ComplianceExtraction(StrictModel):
    document_type: Literal["trade_license", "contract", "iso_cap_certificate", "insurance", "dpa", "regulatory_license", "vendor_agreement", "laboratory_accreditation", "it_software_agreement", "other", "unknown"]
    type_evidence: list[Evidence] = Field(default_factory=list, max_length=5)
    company: TextFact | None = None
    reference_number: TextFact | None = None
    issue_date: DateFact | None = None
    effective_date: DateFact | None = None
    expiry_date: DateFact | None = None
    renewal_date: DateFact | None = None
    termination_notice: NoticeFact | None = None
    parties: list[TextFact] = Field(default_factory=list, max_length=20)
    obligations: list[TextFact] = Field(default_factory=list, max_length=30)
    required_actions: list[Finding] = Field(default_factory=list, max_length=30)
    validation_issues: list[str] = Field(default_factory=list, max_length=30)


class DocumentAnalysis(StrictModel):
    summary: str = Field(max_length=6000)
    summary_evidence: list[Evidence] = Field(min_length=1, max_length=20)
    tasks: list[Finding] = Field(default_factory=list, max_length=100)
    deadlines: list[Finding] = Field(default_factory=list, max_length=100)
    risks: list[Finding] = Field(default_factory=list, max_length=100)
    blockers: list[Finding] = Field(default_factory=list, max_length=100)
    contacts: list[Finding] = Field(default_factory=list, max_length=100)
    expiries: list[ExpiryFinding] = Field(default_factory=list, max_length=100)
    commercials: list[CommercialFinding] = Field(default_factory=list, max_length=100)
    project_status: str | None = None
    requires_attention: bool = False
    compliance: ComplianceExtraction


class ConfidentialTerm(StrictModel):
    value: str = Field(min_length=2, max_length=300)
    kind: Literal["PERSON", "CLIENT", "PROJECT", "ORGANIZATION", "ADDRESS", "VALUE", "CONFIDENTIAL"]


class RulesIn(StrictModel):
    policy: Literal["auto", "review", "test", "skip"] = "auto"
    terms: list[ConfidentialTerm] = Field(default_factory=list, max_length=500)


class ApprovalIn(StrictModel):
    payload_hash: str = Field(pattern=r"^[a-f0-9]{64}$")


class SearchIn(StrictModel):
    q: str = Field(default="", max_length=200)
    cursor: str | None = Field(default=None, max_length=4096)


class ChatMessage(StrictModel):
    role: Literal["user", "assistant", "system"]
    content: str = Field(min_length=1, max_length=4000)


class ChatIn(StrictModel):
    messages: list[ChatMessage] = Field(min_length=1, max_length=50)


class ChatCitation(StrictModel):
    document_id: str
    document_name: str
    document_path: str | None = None
    document_url: str | None = None
    location: str | None = None
    quote: str | None = None


class CentralChatIn(StrictModel):
    messages: list[ChatMessage] = Field(min_length=1, max_length=50)
    document_ids: list[str] | None = Field(default=None, max_length=100)


class ReminderOut(StrictModel):
    id: str
    task_id: str | None = None
    document_id: str
    document_name: str | None = None
    document_path: str | None = None
    document_url: str | None = None
    title: str
    category: str
    target_date: str
    reminder_date: str
    lead_days: int
    responsible_name: str | None = None
    recipient_email: str | None = None
    amount: float | None = None
    currency: str | None = None
    status: str
    notes: str | None = None
    sent_at: str | None = None
    delivery_channels: list[str] | None = None
    last_error: str | None = None
    attempts: int = 0


class ReminderUpdateIn(StrictModel):
    status: Literal["pending", "completed", "dismissed"] | None = None
    target_date: str | None = Field(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$")
    responsible_name: str | None = None
    recipient_email: str | None = None
    notes: str | None = None


class OwnerRuleIn(StrictModel):
    company_id: uuid.UUID | None = None
    document_type: Literal["trade_license", "contract", "iso_cap_certificate", "insurance", "dpa", "regulatory_license", "vendor_agreement", "laboratory_accreditation", "it_software_agreement", "other"] | None = None
    folder_name: str | None = Field(default=None, max_length=128)
    owner_user_id: uuid.UUID | None = None
    owner_department_id: uuid.UUID | None = None
    reminder_leads: list[int] = Field(default_factory=lambda: [60, 30, 28, 21, 14, 7, 6, 5, 4, 3, 2, 1, 0, -1], max_length=100)
    priority: int = Field(default=100, ge=0, le=1000)

    @field_validator("folder_name")
    @classmethod
    def valid_folder_name(cls, value):
        if value is None:
            return None
        value = value.strip()
        if not value or value in {".", ".."} or "/" in value or "\\" in value:
            raise ValueError("Enter one SharePoint folder name, without a path")
        return value

    @field_validator("reminder_leads")
    @classmethod
    def valid_leads(cls, value):
        if any(not -365 <= lead <= 730 for lead in value) or len(value) != len(set(value)):
            raise ValueError("Invalid reminder schedule")
        return value


class ComplianceReviewIn(StrictModel):
    company_id: uuid.UUID
    document_type: Literal["trade_license", "contract", "iso_cap_certificate", "insurance", "dpa", "regulatory_license", "vendor_agreement", "laboratory_accreditation", "it_software_agreement", "other"]
    reference_number: str | None = Field(default=None, max_length=255)
    expiry_date: str | None = None
    renewal_date: str | None = None
    termination_notice_days: int | None = Field(default=None, ge=1, le=730)
    owner_user_id: uuid.UUID | None = None
    owner_department_id: uuid.UUID | None = None
    review_note: str = Field(min_length=3, max_length=2000)

    @field_validator("expiry_date", "renewal_date")
    @classmethod
    def valid_optional_date(cls, value):
        if value is not None:
            from datetime import date
            if date.fromisoformat(value).isoformat() != value:
                raise ValueError("Use an unambiguous ISO date")
        return value


class ComplianceTaskUpdateIn(StrictModel):
    status: Literal["active", "completed"]
    note: str | None = Field(default=None, max_length=2000)


class ComplianceTaskAssignIn(StrictModel):
    owner_user_id: uuid.UUID | None = None
    owner_department_id: uuid.UUID | None = None
    note: str = Field(min_length=3, max_length=2000)

    @field_validator("note")
    @classmethod
    def valid_note(cls, value):
        value = value.strip()
        if len(value) < 3:
            raise ValueError("Explain why the task owner is changing")
        return value

File: frontend/src/api/sharepoint.ts
export type SharePointRun = {
  id: string; status: string; discovered: number; processed: number; failed: number; error_code: string | null;
};

export type SharePointStatus = {
  enabled: boolean; configured: boolean; missing: string[]; connected: boolean;
  microsoft_sign_in_required: boolean; can_review: boolean; user_id: string;
  openai_configured: boolean; active_run: boolean;
  polling_enabled: boolean; sync_interval_seconds: number;
  scheduler_enabled: boolean; email_configured: boolean; teams_configured: boolean;
  run: SharePointRun | null; last_sync: string | null; languages: string[];
};

export type Evidence = { segment_id: string; quote: string };
export type Finding = {
  title: string; owner: string | null; deadline: string | null; status: string; priority: string; evidence: Evidence[];
};

export type ExpiryFinding = {
  title: string;
  date: string;
  category: "expiry" | "renewal" | "effective" | "warranty" | "milestone" | "deadline" | "other";
  responsible: string | null;
  evidence: Evidence[];
};

export type CommercialFinding = {
  description: string;
  amount: number | null;
  currency: string | null;
  payment_terms: string | null;
  billing_frequency: "one_time" | "monthly" | "quarterly" | "annual" | "milestone" | "unknown";
  evidence: Evidence[];
};

export type AnalysisSection = {
  summary: string; summary_evidence: Evidence[]; tasks: Finding[]; deadlines: Finding[];
  risks: Finding[]; blockers: Finding[]; contacts: Finding[];
  expiries?: ExpiryFinding[]; commercials?: CommercialFinding[];
  project_status: string | null; requires_attention: boolean;
  compliance?: {
    document_type: string;
    company?: { value: string } | null;
    reference_number?: { value: string } | null;
    issue_date?: { value: string } | null;
    effective_date?: { value: string } | null;
    expiry_date?: { value: string } | null;
    renewal_date?: { value: string } | null;
    termination_notice?: { days: number } | null;
    parties?: Array<{ value: string }>;
    obligations?: Array<{ value: string }>;
    review_reasons?: string[];
  };
};
export type Segment = { id: string; location: string; text: string };
export type SharePointDocument = {
  id: string; name: string; path?: string; url: string; status: string; error_code: string | null;
  languages: string[]; size?: number; modified_at: string | null; processed_at?: string | null;
  requires_attention: boolean; model?: string; attempts?: number;
  analysis?: { sections: AnalysisSection[]; requires_attention: boolean } | null;
  compliance?: AnalysisSection["compliance"] | null;
  segments?: Segment[]; usage?: { input_tokens: number; output_tokens: number } | null;
};
export type DocumentPage = { items: SharePointDocument[]; next_cursor: string | null };
export type SharePointReminder = {
  id: string;
  task_id?: string | null;
  document_id: string;
  document_name?: string | null;
  document_path?: string | null;
  document_url?: string | null;
  title: string;
  category: string;
  target_date: string;
  reminder_date: string;
  lead_days: number;
  responsible_name: string | null;
  recipient_email: string | null;
  amount: number | null;
  currency: string | null;
  status: "pending" | "sent" | "completed" | "dismissed" | "overdue" | "failed";
  notes?: string | null;
  sent_at?: string | null;
  delivery_channels?: string[] | null;
  last_error?: string | null;
  attempts?: number;
};

export type ChatCitation = {
  document_id: string;
  document_name: string;
  document_path?: string | null;
  document_url?: string | null;
  location?: string | null;
  quote?: string | null;
};

export type ChatMessage = {
  role: "user" | "assistant" | "system";
  content: string;
};

export type ChatResponse = {
  reply: string;
  model: string;
  usage?: { input_tokens: number; output_tokens: number };
  citations?: ChatCitation[];
};

export type CentralChatIn = {
  messages: ChatMessage[];
  document_ids?: string[] | null;
};

export function readableStatus(value: string) {
  const documentErrors: Record<string, string> = {
    analysis_invalid_output: "The AI returned details it could not validate. Retry processing or ask an administrator to review this file.",
    analysis_provider_error: "The AI service could not complete analysis. Please retry later.",
    analysis_rate_limited: "The AI service is busy. Please retry in a few minutes.",
    analysis_auth_error: "The AI connection needs administrator attention.",
    invalid_text_encoding: "This text file could not be read. Save it as UTF-8 or UTF-16 and upload it again.",
    invalid_pdf: "This PDF could not be read. Re-save or repair it, then upload it again.",
    invalid_office_document: "This Word or Excel file could not be read. Re-save it as a valid DOCX or XLSX, then upload it again.",
    unsupported_type: "This file type cannot be analyzed. Supported files are TXT, PDF, DOCX, and XLSX.",
    encrypted_document: "This file is password protected. Upload an unlocked copy for analysis.",
    empty_document: "No readable text was found in this file. Check the original and upload a copy with selectable text.",
    needs_ocr: "This scan has no readable text. Upload a searchable PDF or a clearer copy.",
    ocr_unavailable: "Text recognition is unavailable. Ask an administrator to check the document processor.",
    incomplete_visual_content: "Important content is embedded as images. Upload a searchable copy so it can be analyzed safely.",
    formula_requires_review: "This spreadsheet contains formulas that need review before analysis.",
    text_limit: "This document has too much text to analyze. Split it into smaller files.",
    page_limit: "This PDF exceeds the page limit. Split it into smaller files.",
    image_limit: "This file contains too many or oversized images. Upload a smaller searchable copy.",
    cell_limit: "This spreadsheet has too many cells to analyze. Split it into smaller files.",
    archive_limit: "This Office file exceeds the safe processing limit. Split or simplify it and upload it again.",
    parser_timeout: "Reading this file took too long. Try a smaller or simplified copy.",
    parser_memory_limit: "Reading this file exceeded the processor memory limit. Try a smaller copy.",
    file_too_large: "This file is too large to process. Upload a smaller copy.",
    document_changed_sync_required: "This file changed in SharePoint during processing. Sync again to analyze the latest version.",
    analysis_incomplete_or_refused: "The AI did not finish analyzing this file. Retry processing or ask an administrator to review it.",
    openai_not_configured: "AI analysis is not configured. Ask an administrator to check the connection.",
    extraction_failed: "The file could not be read. Check that it opens correctly, then upload it again.",
  };
  if (documentErrors[value]) return documentErrors[value];
  return value.replace(/_/g, " ");
}

export function formatBytes(bytes?: number): string {
  if (!bytes || bytes <= 0) return "—";
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

export function getStatusBadgeInfo(status: string): {
  label: string;
  variant: "default" | "secondary" | "destructive" | "outline";
  className?: string;
} {
  switch (status) {
    case "ready":
      return { label: "Ready", variant: "default", className: "bg-emerald-600/15 text-emerald-800 dark:text-emerald-300 border-emerald-600/30 font-medium" };
    case "processing":
      return { label: "AI Processing…", variant: "outline", className: "bg-primary/10 text-primary border-primary/30 animate-pulse font-medium" };
    case "queued":
      return { label: "Queued", variant: "outline" };
    case "failed":
      return { label: "Failed", variant: "destructive" };
    case "ai_skipped":
      return { label: "AI Prohibited", variant: "outline" };
    default:
      return { label: readableStatus(status), variant: "outline" };
  }
}

export function calculateEstimatedCost(
  usage?: { input_tokens: number; output_tokens: number } | null,
  modelName?: string | null
): { costUsd: number; formatted: string } | null {
  if (!usage) return null;
  const model = (modelName || "gpt-5.6-luna").toLowerCase();
  let inRate = 0.20 / 1_000_000;
  let outRate = 1.20 / 1_000_000;

  if (model.includes("gpt-5.6-luna")) {
    inRate = 0.20 / 1_000_000;
    outRate = 1.20 / 1_000_000;
  } else if (model.includes("gpt-5.6-sol")) {
    inRate = 4.00 / 1_000_000;
    outRate = 20.00 / 1_000_000;
  } else if (model.includes("gpt-5.6-terra")) {
    inRate = 2.00 / 1_000_000;
    outRate = 12.00 / 1_000_000;
  } else if (model.includes("gpt-4o-mini")) {
    inRate = 0.15 / 1_000_000;
    outRate = 0.60 / 1_000_000;
  } else if (model.includes("gpt-4o")) {
    inRate = 2.50 / 1_000_000;
    outRate = 10.00 / 1_000_000;
  } else if (model.includes("o3-mini") || model.includes("o4-mini")) {
    inRate = 1.10 / 1_000_000;
    outRate = 4.40 / 1_000_000;
  } else if (model.includes("o1")) {
    inRate = 15.00 / 1_000_000;
    outRate = 60.00 / 1_000_000;
  }

  const cost = (usage.input_tokens * inRate) + (usage.output_tokens * outRate);
  const formatted = cost > 0 && cost < 0.01
    ? "<$0.01"
    : `$${cost.toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
  return { costUsd: cost, formatted };
}

export function formatDateTime(iso?: string | null): string {
  if (!iso) return "—";
  try {
    const d = new Date(iso);
    if (isNaN(d.getTime())) return iso;
    return d.toLocaleString(undefined, {
      month: "short",
      day: "numeric",
      year: "numeric",
      hour: "2-digit",
      minute: "2-digit",
    });
  } catch {
    return iso;
  }
}
